File: 1erdepartamental/python/7.creacion_dato.py
# ============================================
# PROGRAMA: Creación de Tipos de Datos
# DESCRIPCIÓN: Modela entidades del mundo real con POO
# ============================================


class Persona:
    """
    Representa una persona con encapsulación completa.

    __str__ define cómo se imprime el objeto
    """

    def __init__(self, nombre, ap, am, genero, edad):
        self.__nombre = nombre
        self.__ap = ap
        self.__am = am
        self.__genero = genero
        self.__edad = 0
        self.edad = edad  # Usa setter para validación

    @property
    def nombre(self):
        return self.__nombre

    @property
    def edad(self):
        return self.__edad

    @edad.setter
    def edad(self, valor):
        if valor < 0:
            raise ValueError("La edad no puede ser negativa")
        self.__edad = valor

    def nombre_completo(self):
        return f"{self.__nombre} {self.__ap} {self.__am}"

    def __str__(self):
        """Representación legible para print()"""
        return f"{self.nombre_completo()}, {self.__edad} años"

    def __repr__(self):
        """Representación técnica para debugging"""
        return f"Persona('{self.__nombre}', '{self.__ap}', '{self.__am}', '{self.__genero}', {self.__edad})"


class Auto:
    """Representa un automóvil con formateo de precio"""

    def __init__(self, marca, modelo, precio, anio):
        self.__marca = marca
        self.__modelo = modelo
        self.__precio = 0
        self.__anio = anio
        self.precio = precio  # Usa setter

    @property
    def precio(self):
        return self.__precio

    @precio.setter
    def precio(self, valor):
        if valor < 0:
            raise ValueError("El precio no puede ser negativo")
        self.__precio = valor

    def __str__(self):
        """Formatea precio con separadores de miles"""
        return f"{self.__marca} {self.__modelo} ({self.__anio}) - ${self.__precio:,.2f}"

    def __repr__(self):
        return (
            f"Auto('{self.__marca}', '{self.__modelo}', {self.__precio}, {self.__anio})"
        )


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    p = Persona("Hugo", "Dominguez", "Lopez", "M", 19)
    print(p)
    print(repr(p))

    print()

    a = Auto("Honda", "Civic", 350000, 2022)
    print(a)
    print(repr(a))

File: 1erdepartamental/c++/7.creacion_dato.cpp
// ============================================
// PROGRAMA: Creación de Tipos de Datos
// DESCRIPCIÓN: Modela entidades con POO
// ============================================

#include <iostream>
#include <string>
#include <stdexcept>
using namespace std;


/**
 * Representa una persona con encapsulación completa
 */
class Persona {
private:
    string __nombre;
    string __ap;
    string __am;
    char __genero;
    int __edad;
    
    void validarEdad(int edad) const {
        if (edad < 0) {
            throw invalid_argument("La edad no puede ser negativa");
        }
    }

public:
    Persona(string nombre, string ap, string am, char genero, int edad)
        : __nombre(nombre), __ap(ap), __am(am), __genero(genero), __edad(0) {
        setEdad(edad);  // Usa setter para validación
    }
    
    // Getters
    string getNombre() const { return __nombre; }
    int getEdad() const { return __edad; }
    
    // Setter con validación
    void setEdad(int edad) {
        validarEdad(edad);
        __edad = edad;
    }
    
    string nombreCompleto() const {
        return __nombre + " " + __ap + " " + __am;
    }
    
    // Operador << para imprimir
    friend ostream& operator<<(ostream& os, const Persona& p) {
        os << p.nombreCompleto() << ", " << p.__edad << " años";
        return os;
    }
};


/**
 * Representa un auto con formateo
 */
class Auto {
private:
    string __marca;
    string __modelo;
    double __precio;
    int __anio;
    
    void validarPrecio(double precio) const {
        if (precio < 0) {
            throw invalid_argument("El precio no puede ser negativo");
        }
    }

public:
    Auto(string marca, string modelo, double precio, int anio)
        : __marca(marca), __modelo(modelo), __precio(0), __anio(anio) {
        setPrecio(precio);
    }
    
    double getPrecio() const { return __precio; }
    
    void setPrecio(double precio) {
        validarPrecio(precio);
        __precio = precio;
    }
    
    friend ostream& operator<<(ostream& os, const Auto& a) {
        os << a.__marca << " " << a.__modelo << " (" << a.__anio << ") - $";
        os.precision(2);
        os << fixed << a.__precio;
        return os;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    Persona p("Hugo", "Dominguez", "Lopez", 'M', 19);
    cout << p << endl << endl;
    
    Auto a("Honda", "Civic", 350000, 2022);
    cout << a << endl;
    
    return 0;
}
